vllm/inputs/data.py

```diff
@@ -122,6 +123,12 @@ class LLMInputs(TypedDict):
     if the model supports it.
     """
 
+    token_annotations: NotRequired[Optional[List["TokenRangeAnnotation"]]]
+    """
+    Optional token annotations to capture content that will replace portions
+    of the token IDs list.
+    """
+
```

Comment:
Given how #8346 has evolved (placeholder ranges are now propagated instead of inlined into MM data) I'll likely remove this and instead compute the annotations downstream once that change lands. But the rest of the change should still be applicable :)

Reply:
Just want to clarify, what is the difference between the placeholder range data structures in #8346 and the `TokenRangeAnnotation` data structures in this PR?

Given that you are not creating any new example scripts in this PR, am I correct that the placeholder range data structures in #8346 are more "frontend-oriented" and serve to align placeholder tokens with multimodal input within the prompt (in a way that is model- and workload-specific), while the TokenRangeAnnotation data structures in this PR are more "backend-oriented" and serve to bridge multimodal data into core engine functionality such as prefix cache, block management, etc? With the idea being that the TokenRangeAnnotation's will be computed from the placeholder token range data structures?

Reply:
Yup, that frontend/backend distinction was how I was thinking about it, but I could be convinced to combine them too.